[RFC] memline: fix segfaults

justinmk opened the pull request:
- regression by 42f1bd9
- closes #1836
- fixed another segfault from  00f18e2e09af35494d0b94ae76a9dbc3e0319d89
- ported legacy test  78, which is related to swapfile/recovery

todo:
- [x] delete old test78
- [ ] another (no. 3) segfault: https://github.com/neovim/neovim/issues/2019#issuecomment-75370337
- [ ] E309: https://github.com/neovim/neovim/pull/1805#issuecomment-91271069

justinmk committed:
passing-by: mark non-null return; remove invalid OOM check.

remove some #ifdef noise

justinmk committed:
fix: segfault in recover_names(): must not concat full path

- regression by 42f1bd9b2228aaca4fb8a5597a3b5774f7ef6876
- closes #1836

aktau commented:
LGTM, nice catch :). It clearly looks like those two lines after the `else` belong(ed) together and possibly got mixed up by some automatic tool. Strange.

oni-link commented on src/nvim/memline.c:
```diff
@@ -1571,16 +1552,10 @@ static time_t swapfile_info(char_u *fname)
 
 static int recov_file_names(char_u **names, char_u *path, int prepend_dot)
 {
```
We could add `FUNC_ATTR_NONNULL_ALL`.

justinmk replied:
:+1:

jszakmeister commented:
*nudge*  Is there anything left to do here?  Just the `FUNC_ATTR_NONNULL_ALL` bit?

justinmk committed:
test: migrate legacy test 78

Pyrohh commented:
Is the core of this change just the missing braces? If the other commits aren't needed or ready (given the `WIP` in one of the names). Then perhaps just that braces change could be commited.

philix commented:
@aktau if the preprocessor test didn't succeed (almost never the case) the two lines inside a brace were just a normal block that didn't belong to an `else`.

I probably removed them thinking they were a leftover of some of my refactorings that simply removed the `if (out of memory)` text leaving the `{` block untouched.

It turns out that leaving orphan blocks was a safer strategy, but I started removing them after some negative review feedback. Removed too much though :)

@justinmk that's an incredible catch. Thanks for fixing and sorry for introducing it.

philix commented:
Illustration for my comment above:
![screen shot 2015-03-31 at 12 50 53 am](https://cloud.githubusercontent.com/assets/207795/6912026/03f7c14a-d740-11e4-924e-53c9c74445f9.png)

justinmk committed:
fix: segfault in ml_recover() if swap file not found

- To reproduce: ":recover non_existent_swapfile"
- Regression: 00f18e2e09af35494d0b94ae76a9dbc3e0319d89

justinmk commented on src/nvim/memline.c:
```diff
@@ -1253,8 +1253,10 @@ void ml_recover(void)
   }
-  free(buf->b_ml.ml_stack);
-  free(buf);
+  if (buf != NULL) { //may be NULL if swap file not found.
```
Segfault occurs because fname_used may be null on line 836, then `buf` is never initialized.

Pyrohh replied:
Should be two spaces after `{`.

I think it might be a good idea to clean up files so we can take them off of `clint-ignored-files.txt`, but of course later.

justinmk commented on src/nvim/memline.c:
```diff
@@ -1330,53 +1332,35 @@ recover_names (
-        num_names = 0;
-        break;
-      }
-    }
```
Also removing this OOM check.

justinmk commented:
Reason this PR was held up is because I was trying to write a test, and decided to port test 78 since it is related, and ran into another segfault, plus some unrelated issues.

justinmk commented:
You can reproduce the other segfault like this:

    :recover foo

Since foo isn't an existing swapfile, it will crash. Reason this wasn't caught by the tests is that the tests always work with actually existing swapfiles.

justinmk committed:
test regression: recover_spec.lua

justinmk commented:
This is ready for review if anyone wants to comment. I'm going to add one more test, then merge.

justinmk committed:
test: helpers.lua: add temporary set_session() function

This function is needed until helpers.lua is refactored to support
multiple sessions.

justinmk commented on test/functional/helpers.lua:
```diff
@@ -49,6 +53,10 @@ end
 
+local function set_session(s)
+  session = s
+end
```
@bfredl @tarruda This is a temporary hack to allow multiple spawned sessions in a single test. After this PR I plan to refactor helpers.lua into a module with a `new()` constructore so that a test can control multiple sessions. Does that sound ok?

Another thing that would be useful (for testing swapfiles) is ability to `kill -9` the session. Any nice way to do that? I see that `loop.c` in the lua client [does not send signals](https://github.com/neovim/lua-client/blob/master/nvim/loop.c#L58).

tarruda replied:
> After this PR I plan to refactor helpers.lua into a module with a new() constructor so that a test can control multiple sessions. Does that sound ok?

:+1: 

> Another thing that would be useful (for testing swapfiles) is ability to kill -9 the session. Any nice way to do that? I see that loop.c in the lua client does not send signals.

I added session:exit() to accept arguments for sending SIGTERM/SIGKILL: https://github.com/neovim/lua-client/blob/master/nvim/loop.c#L154-L156, but forgot to forward parameters from the session layer to the loop layer: https://github.com/neovim/lua-client/blob/master/nvim/session.lua#L127-L129 . It should be very simple to enable this functionality, could you send a PR to lua client?

justinmk replied:
Sure, I wasn't sure if it was intentional.

justinmk commented on test/functional/ex_cmds/recover_spec.lua:
```diff
@@ -0,0 +1,71 @@
+    --TODO(justinmk): this is an ugly hack to force `helpers` to support
+    --multiple sessions.
+    local nvim2 = helpers.spawn({helpers.nvim_prog, '-u', 'NONE', '--embed'})
+    helpers.set_session(nvim2)
```
To control a session session in the same test, use the `set_session()` function. All the calls to helper functions in the rest of this test refer to that second session. This is pretty ugly, so I plan to refactor helpers.lua shortly.

tarruda replied:
It's my fault, I should have wrapped session methods into another class for the tests :disappointed:
